curated_transformers/pipeline/distill.py

```diff
@@ -118,37 +127,49 @@ def distill(
             return losses
 
         student_hidden, student_backprop = self.model.begin_update(student_docs)
-
         # We have to pool the teacher output in the same way as the student.
         student_pooler = self.model.get_ref("tok2vec").get_ref("pooling")
-
         teacher_hidden = with_ragged_layers(student_pooler).predict(teacher_docs)
-
         layer_mapping = self._layer_mapping(teacher_hidden, student_hidden)
 
-        normalize = True
-        d_mse = []
-        for t_doc, s_doc in zip(teacher_hidden, student_hidden):
-            if normalize:
-                d_mse_doc = [
-                    (s_doc[student_layer] - t_doc[teacher_layer])
-                    / self.model.ops.xp.linalg.norm(t_doc[teacher_layer])
-                    for student_layer, teacher_layer in layer_mapping
-                ]
-            else:
-                d_mse_doc = [
-                    s_doc[student_layer] - t_doc[teacher_layer]
-                    for student_layer, teacher_layer in layer_mapping
-                ]
-
-            d_mse.append(d_mse_doc)
-
-        losses[self.name] += sum(
-            sum(float((d_layer**2).sum()) for d_layer in d) for d in d_mse
-        )
-
-        student_backprop(d_mse)
-        if sgd not in (None, False):
+        # Transpose the representations to the get the following shape: [layer, batch, seq, repr]
+        t_student_hidden = list(zip(*student_hidden))
+        t_teacher_hidden = list(zip(*teacher_hidden))
+
+        # Allocate loss accumulators on the GPU if possible.
+        cum_loss = self.model.ops.alloc1f(1)
+        layer_loss = self.model.ops.alloc1f(1)
+        t_mses = []
+        for student_layer_idx, teacher_layer_idx in layer_mapping:
+            student_batches = t_student_hidden[student_layer_idx]
+            teacher_batches = t_teacher_hidden[teacher_layer_idx]
+
+            batch_size = len(student_batches)
+            n_elements = student_batches[0].size
+            layer_loss.fill(0)
+            batch_mses = []
+            for s_doc, t_doc in zip(student_batches, teacher_batches):
+                # Calculate MSE loss between student and teacher representations.
+                mse = (s_doc - t_doc) ** 2
+                if self.loss_normalization == "mean":
+                    layer_loss += mse.sum()
+                elif self.loss_normalization == "squared_l2_norm":
+                    norm = self.model.ops.xp.linalg.norm(t_doc.reshape(-1)) ** 2
+                    mse /= norm
+                    layer_loss += mse.sum()
+                batch_mses.append(mse)
+
+            t_mses.append(batch_mses)
+            if self.loss_normalization == "mean":
+                cum_loss += layer_loss / (batch_size * n_elements)
```

Comment:
I have to recheck the derivatives, but I with mean loss, the gradients should also be scaled. Empirical example:

```
>>> X = torch.full((10,), 0.5, requires_grad=True)
>>> loss = ((X - torch.ones(10)) ** 2).mean()
>>> loss.backward()
>>> X.grad
tensor([-0.1000, -0.1000, -0.1000, -0.1000, -0.1000, -0.1000, -0.1000, -0.1000,
        -0.1000, -0.1000])
```

The code above would result in a gradient of `-0.5` for all parameters.

Maybe it's worth lifting the loss/gradient calculation out of this loop, to that we can test against Torch in a unit test? To easy to get it wrong...

Reply:
I've refactored the loss calculation into its own class and added unit tests.